### ISSUE TITLE
alternate PAKE algorithms

### ISSUE THREAD

**Issue by warner** [warner/magic-wormhole#348]
I was chatting with @gtank and @str4d about alternate PAKE algorithms the other week (in particular SPAKE2-EE over Ristretto, basically the blinding factor is the Elligator2 string->point mapping of the password, instead of hash-password-to-scalar and multiplying by hash-constant-to-point). I figured I should write down some notes about what room we have in the protocol to negotiate the PAKE phase, and what effects multiple algorithms might have on security.

The most privileged position an attacker might achieve, within our threat model, is control over the "mailbox server", which is the centralized machine that helps get key-exchange messages from one client to the other (and later, it helps with other messages encrypted by the negotiated session key). The attacker might do this by compromising this server, or stealing the credentials of its administrator, or being the administrator themselves (i.e. me). Until we get #144 implemented, they could also camp out on the server's network connection and manipulate the unencrypted TCP traffic over which the mailbox protocol currently runs. Or they could camp out in front of a single user's connection and control all traffic from that one user.

## Protocol

For background, here is a summary of the magic-wormhole server, as visible to the mailbox server.

First, the sender establishes a Websocket connection to the server and asks to allocate a "nameplate" (the numerical prefix in the wormhole code) and a "mailbox" (a large random string, associated with an initially-empty set of messages). Then the sender chooses the random portion of the PAKE code (the words) and generates a PAKE message, depositing it into the mailbox. They tell the recipient about the full wormhole code (nameplate plus random words) and wait for a new message to appear in the mailbox.

The recipient creates the Websocket connection to the server immediately upon program launch. As the user types in the wormhole code, once the initial nameplate segment is entered, the program sends a "CLAIM" message to exchange the nameplate for a mailbox identifier. At that point, neither side needs the nameplate anymore, and it is released for use by other connections (perhaps subject to a delay, see #31). The sender's PAKE message is fetched.

A few seconds later, the receiving user finishes typing in the rest of the code, and the recipient machine's PAKE message is created, and submitted to the server. The recipient computes the PAKE session key, and creates the "VERSION" message: this is a bundle of feature identifiers, serialized, and encrypted by the session key. (To be precise, the serialized string is put through the libsodium "SecretBox" authenticated encryption function, using a key derived via HKDF from the session key, a "phase identifier" of `version`, and a random "side identifier" selected by each program when it launches and attached to all messages). The encrypted VERSION messge is submitted to the server, where it is added to the mailbox along with the PAKE message.

The sender is notified of the recipient's PAKE message over the Websocket connection. It computes the shared PAKE key, then generates+submits its own VERSION message. It then hears about the receipient's VERSION message and attempts to authenticated+decrypt it. If this succeeds, we know it was produced by someone who knew the correct PAKE code. The sender's application is informed of a verified connection, at which point it will proceed with a file transfer, effectively encrypting the file with the session key (again, more key derivation is involved).

The recipient gets the sender's encrypted VERSION message and does the same. In this way, the VERSION message serves double duty as a "Key Confirmation Message", the "KCM".

## Attacks

Magic-Wormhole's promise is to "get things from one computer to another, safely". There are three things the attacker might usefully try to accomplish that would violate this (not including simple disruption of the connection, which would certainly be annoying but is outside the threat model):

* 1: violate confidentiality: learn the contents of the file and/or message being sent
* 2: violate integrity: cause the recipient to get something other than what the sender intended
* 3: both: learn the contents of the file *and* control what gets delivered to the recipient

The first attack "wins" if the attacker gets to see the file. It's still a successful attack even if the recipient never gets to see the file too (they might get an error, or a timeout). And it's successful even if either party discovers the attack, as long as the sender discovers it too late to avoid transmitting the data.

The second attack is successful if the attacker can deliver a bogus file (think malware, with the positive reputation of having come from the intended sender). The attacker doesn't have to see the original file first, so the sender might get an error (and be alerted to the attack), but the attacker prevents the recipient from hearing their warnings until it's too late.

There is a weaker form, in which the attacker might manage to undetectably corrupt the file, but doesn't get complete control over their modifications (think bit-flipping attacks in a non-authenticated cipher mode). The AEAD we use should prevent these.

The third attack, if successful, allows the attacker to examine the transmitted file before committing to the recipient's version. This attack lets them deliver a modified version of the genuine file, which might make the malware attack more interesting.

If the wormhole connection is long-lived and used for more than a single file transfer, the three attacks correspond to:

* 1: The sender is convinced they're talking to the recipient, but actually they're talking to the attacker. The recipient knows they aren't talking to anyone.
* 2: The sender knows they aren't talking to anyone. The recipient thinks they're talking to the sender, but they're actually talking to the attacker.
* 3: Both sender and recipient think they're talking to each other, but in fact they're both talking to the attacker, who might be passively relaying everything, or might be inventing both conversations on the fly.

The third attack is the most difficult. The other two are roughly twice as easy.

Assuming the PAKE algorithm is sound, and all the usual other assumptions hold true (the RNG is uniform, there are no useful timing channels to be had, the implementations are correct), the primary threat is the short wormhole code. The default `--code-length=2` and the 256-entry wordlist gives us 16 bits of entropy per code. The codes are selected (uniformly) randomly for each invocation of the program.

An attacker who wants to achieve "3" (full MitM) gets one chance to guess the code correctly. Since the codes are random, their strategy is simple: guess `adroitness-aardvark` (the first words in the lists) each time, and hope they got it right (they'll win with probability `2^-16`). The attacker, controlling the server, will perform the protocol with just the sender, although to reduce suspicion they might wait until the recipient connects and submits their messages too. The attacker sees the sender's PAKE message, chooses their guess, and emits their own PAKE message. At this point, they're committed to a single guess. The sender will see their PAKE, compute the shared key, send their encrypted VERSION message (the KCM), and wait for the recipient's VERSION/KCM message. The attacker will perform trial decryption of the KCM. If it succeeds, the attacker knows they've won, and they can now perform the same dance with the recipient, knowing they'll win there too.

If it fails, the attacker learns about the mismatch before the sender does. The sender won't notice until it sees an encrypted VERSION/KCM message that fails to authenticate, and the attacker gains nothing by sending a KCM that it knows will be detected as a fraud. At this point, the attacker's best strategy is to stay quiet forever, or to drop the connection, and hope the victims will think they have an unreliable network, and try again. This gives the attacker another chance, one per execution of the protocol, and is mainly limited by how quickly the victims get frustrated and give up (unlikely more than a handful of times).

The sender might display a warning if a PAKE message is not quickly followed by a VERSION/KCM message, but there are legitimate configurations which could cause this delay, so it isn't clear that a security warning would be useful.

The recipient, still waiting, sees no messages at all until the attacker knows they have won.

If the attacker is content to merely achieve goals 1 or 2 (a single-sided compromise rather than a full monster-in-the-middle), then they get two chances to guess. Suppose their goal is to deliver a bogus file to the recipient. They perform the above attack on the sender, and use the sender's KCM to tell whether they succeeded or not. If so, great, use `adroitness-aardvark` with the recipient and expect success. If the sender's KCM indicated a failure, however, they use `adviser-aardvark` with the recipient (the second possible code). This is equally likely to fail, but equally likely to succeed.

If the first guess worked, the attacker gets a full MitM win (goal 3). If the first guess fails but the second works, the sender sees a network failure, but the recipient sees a legitimate-looking connection (goal 2). If the second guess fails too, the recipient sees the protocol stall halfway (they receive the PAKE but not the KCM), just like the sender did.

So the attacker gets a `2*2^-16` chance of goal 2. If the attacker cared more about seeing the file's contents (goal 1) than controlling them (goal 2), they can flip the attack: wait until the receiver connects and perform their first attack on the receiver, then use the results of that attack to guide their guess for the sender.

## Multiple PAKE Algorithms

Suppose we want to roll out a new PAKE algorithm, perhaps because it is faster, or has better security properties. We have a couple of different strategies to choose from:

* negotiate the algorithm first: before computing any PAKE values, each side publishes (and waits for) a messge with a list of the algorithms they understand. Then either they decide on a leader (highest "side" string wins) and the follower waits to hear the leader's decision, or they use convergent algorithms to independently decide upon the same one. Then they publish their PAKE messages and proceed as before.
* make multiple offers in parallel: compute the PAKE message for every algorithm they know, send them all together in the same server message. Use some form of negotiation to decide which one to respond to. Respond to only one.

The "respond to only one" is the important part, for security at least. The attacker gets one guess with the sender, and one guess with the recipient, and if their first guess is wrong they can't get the full both-sided compromise, and if their second guess is wrong too then they don't get *any* compromise.

The downside of making multiple offers in parallel is the CPU time spent creating PAKE messages that will never be used. But the upside is that it avoids the round trip needed to negotiate the algorithm first, plus the excess computation can be hidden in the latency (the intense CPU work is performed while waiting for the recipient to connect, although on modern machines this only takes a fraction of a second).

The second participant (usually the recipient) learns the shared algorithm decision first (as soon as the nameplate/mailbox is opened and the first messages arrive, which can happen before the user has finished typing in the remainder of the code if their network is fast), so they only have to compute one set of parameters. This computation is on the critical path, however, so they can't hide it in the latency like the sender can.

For security, the thing to avoid is having either side compute the second half of the PAKE computation more than once per code, and revealing the results to the attacker (via the KCM message). If both sides generated PAKE messages for two algorithms, and also generated KCM messages for both, the attacker would have four opportunities to test their guess, not just two.

## Where To Add This To The Protocol

The PAKE message sent to the server is a JSON-serialized dictionary built in the `build_pake()` function of `_key.py`. It currently contains a single key, `pake_v1`, whose value is a hex-encoded bytestring of the output of the `start()` method of SPAKE2 as implemented in python-spake2, using the `Ed25519` group in the "Symmetric" mode:

https://github.com/warner/magic-wormhole/blob/ab5fe65c3beee5d8cc0d3334beaf0bbac3418c32/src/wormhole/_key.py#L187-L193

To add a second algorithm, we should extend this message. It should include a `pake_v2` or `pake_ALGORITHM_NAME_vNN` field. In `got_pake()` and `compute_key()` of the same file, we must perform the implicit version negotiation. Each PAKE message includes the "side" string, which can be compared to figure out who is the leader, if that matters. However we must either add another message for the leader to express their decision (adding a roundtrip), or use convergent algorithms so both sides can come to the same decision without additional messages.

In the Rust port, it lives inside `KeyMachine.start`:

https://github.com/warner/magic-wormhole.rs/blob/3a9055432a4174da78b0b952f23933ad8cea4b8f/src/core/key.rs#L54-L63

**Comment by warner** [comment]
Oh, and we were thinking that the alternate PAKE algorithms could be added to the API as an optional parameter: `w = wormhole.create(appid, ... pakes=[V1, V2])`, with the default `=[V1]` being the existing Ed25519/SPAKE2 scheme. We might want to combine the PAKE selection with the symmetric encryption scheme, ala TLS ciphersuites, so the parameter name might want to be different.

As far as compatibility goes.. my main goal is making sure that the existing deployments of `/usr/bin/wormhole` continue to interoperate with newer ones. It's been distributed in debian/ubuntu LTS releases by now, so there's a nonzero number of older clients who won't know how to speak a new PAKE scheme. So I want the existing Ed25519-based scheme to be available for a while.

But, if you're writing a new application (new APPID), then the only thing you need to be compatible with is yourself. So adding new options is fine. And if you were porting MW to a different language entirely, then as long as you don't care about writing a `/usr/bin/wormhole`-compatible file-transfer tool in that language, you don't need to implement the Ed25519 PAKE scheme at all. We should of course coordinate on IDs for the different schemes so that two non-interoperable programs *know* that they aren't interoperable, rather than having them disagree about what `pake_v2` means.

**Comment by warner** [comment]
It occurred to me today that the same approach I described for doing multiple PAKE algorithms without incurring a roundtrip, would also let you use a "sided" algorithm without adding roundtrips.

We're currently using a slightly-strange form of SPAKE2 in which we set the two blinding elements to be equal to each other (`M == N`, or `U == V`, depending upon which paper you're reading). Mike Hamburg told me he though it would be equally secure, but the proof would probably bottom out in a different set of assumptions (one of the various DH hardness assumptions). I wanted M==N because that means the two sides don't have to agree upon which side they're playing beforehand. This reduces the number of roundtrips, especially for the "offline codes" scenario, in which Alice and Bob first meet without their computers, agree upon a code, then walk home and type the codes into their machines (`wormhole send --code XYZ file.jpg` and `wormhole receive XYZ`). In the "connect" mode (that will be enabled by Dilation), they don't even agree to a sender/receiver role: they both do the same operation (`wormhole connect --code XYZ`). There's nothing to distinguish them until they can exchange a randomly-generated "side" parameter and then compare to see whose is greater. So it's pretty helpful if the PAKE protocol doesn't force them into picking a side.

The way to avoid the roundtrip while sticking to `M != N` is to have both sides create both PAKE messages: they run the protocol twice with the same password, but once as side A and again as side B (each execution gets a different random DH scalar, of course). They put both PAKE messages into the initial payload, marked as "this is the PAKE message for when I'm being A" and "this is the PAKE message for when I'm being B", along with the random "side" value.

When the client receives their peer's pair of PAKEs, they compare sides and figure out who will be A and who will be B. They then use one of the messages and ignore the other. By sticking to the rule that you only ever complete the second half of the PAKE protocol once (per password), we don't lose any security ground. The cost is a doubling of the CPU work to build the initial message (about 3.8ms on a modern CPU in Python, 0.1ms in Rust, 180ms in Python on a Raspberry Pi Zero). But the benefit is that we stay closer to the security proofs.

**Comment by mentalisttraceur** [comment]
Question: hat was the reason for choosing SPAKE2? In particular, I am curious if you consider it better than CPace?

(This seemed like a decent place to ask, since 1: this is the first issue I found when I looked to see if this was already asked; and 2: seems relevant as motivation for "how much should a contributor/user with limited time worry about the current PAKE choice or helping change it?")

**Comment by mentalisttraceur** [comment]
Ah, I found [your review from the PAKE standardization competition](https://mailarchive.ietf.org/arch/msg/cfrg/BBQ2gwCECu5ouTJjE_CE6d9Rg-0/). I have started reading it, perhaps the answer is in there.

**Comment by mentalisttraceur** [comment]
Okay I see. If I understood right, reducing round trips seemed to be the big factor, helped by SPAKE2 being simpler to implement and more concretely/fully specified.

**Comment by warner** [comment]
Yeah, I picked SPAKE2 well before that competition happened (or at least before I became aware of it). At the time I was familiar with J-PAKE, having used it on the first version of Firefox Sync, and when I explained our needs to Dan Boneh at a conference, he recommended SPAKE2 instead, and said it was described in his book (https://toc.cryptobook.us/) under the name of "PAKE2" (looks like chapter 21.11, page 909, of the current version: https://crypto.stanford.edu/~dabo/cryptobook/BonehShoup_0_6.pdf).

And yeah, magic-wormhole specifically benefits from a symmetric algorithm, whereas most of the competition entries were addressing an asymmetric use case of a web server managing a set of users logging in. It was a long time ago (I'd have to reread my review comments), I remember the phrase "augmented PAKE" being involved, but I think it was possible to use one form as the other but at the cost of an extra message.

It was vaguely my intention to look at the winner of that competition and change magic-wormhole to use it (assuming the flavor worked for us), but I haven't heard of any problems with SPAKE2, it's pretty simple to explain and implement, and the backwards-compatibility effort probably wouldn't be worth any potential benefits.

**Comment by cquike** [comment]
I wonder whether moving away from SPAKE2 would help also with  #405 , since the spake2 library used  https://github.com/warner/python-spake2/blob/master/src/spake2/spake2.py seems to be unmaintained. Or if SPAKE2 is still used whether a different library which is well supported can be used.

**Comment by cquike** [comment]
No, I am afraid I couldn't find an alternative SPAKE2 library....

**Comment by hondogitsune** [comment]
```https://github.com/stef/libopaque
https://github.com/cfrg/pake-selection
https://bsky.app/profile/filippo.abyssdomain.expert/post/3ljcpfkurut2y
https://github.com/BjoernMHaase/AuCPace/blob/master/python/aucpace_testvectors.ipynb```

Admitting all winners are still drafts due to lots of red tape, libopaque seems promising. 

As Victor Shoup presented at TCC 2020, SPAKE2 is not UC secure:
```https://iacr.org/cryptodb/data/paper.php?pubkey=30638```

@warner 
Would this year be good to implement a draft version as optional alternative PAKE? Something to have once finalized, but to play with in the meantime. PAKE2 was good, but it aged. Of course I'm not a cryptographer, but I want to revive the dialogue about this.

**Comment by meejah** [comment]
I'm certainly open to having additional PAKE algorithms as options -- especially if they're "opt in" or otherwise experimental, etc.

One also has to be careful about putting "negotiation" into the protocol, so the easiest might be something like you suggest and have the server operator decide.

At least, I think that's what you mean with the CLI argument suggestion -- i.e. that the server would specify `--use-opaque` (or maybe that's `--pake-algorithm=opaque` or something) and clients have to use whatever the server says?

I am also not a cryptographer ;)

**Comment by hondogitsune** [comment]
@meejah 
Yes purely optional for testing. My reasoning was:
1) Could be beneficial if augmented
2) Should be a winner of round 2
3) Should have an IETF draft and a IACR publication
4) Should have a python implementation already at hand

For the PQ requirement:
```OPAQUE may also be instantiated with any post-quantum (PQ) AKE protocol```

If you remember from your other endeavours in cryptography. ML-KEM is already rolled out and makes up for almost all Cloudflare TLS1.3 traffic. This is just an example, unrelated to this PAKE discussion, but PQ compatibility is a very good bonus.

CFRG being slow, ristretto taking 5 years and EdDSA only becoming FIPS approved in 2023 when it was already year in use by cryptography enthusiasts is merely a reminded that waiting not always is beneficial. Today Ed25519 and Ristretto, Decaf and Elligator are used almost anywhere.

I'd be also interested in the other finalist AuCPace, but I cannot find a real implementation. Just the Jupyter Notes example that is absolutely far away from any usable python library. Hence yes, an optional "opt-in" via switch for ```--pake-algorithm=opaque``` as you suggested would be much welcome.

If I was more competent I would make a pull request, but I don't understand the codebase well enough to integrate anything.  

As someone once suggested how to evaluate such propositions:
What are the benefits:
Have an approved round 2 finalist ready as the draft is being finalized in parallel.
What are the harms:
Low, as it remains optional. A small line in the output could remind the user this is only for testing.
How hard is it to implement:
I cannot answer. Probably intermediate if the library integrates well. Only warner would know.

**Comment by meejah** [comment]
Yes, sometimes cryptography "standards" move slowly ;)

Probably "OPAQUE" because I've at least read about that one before, and it's got an RFC and everything.

There is https://pypi.org/project/opaque/ which wraps a C implementation (https://github.com/stef/libopaque/). There's also a Rust implementation (https://github.com/facebook/opaque-ke) which makes me fell better than C -- but I've not evaluated any of those (and not sure if there are Python bindings for the Rust thing).

Opinions welcome on any of these :)

**Comment by hondogitsune** [comment]
Well libopaque has their own Python bindings made at home, by the lead dev, and to not dismiss opaque-ke in Rust I tried to find some Python bindings for it as well: https://github.com/tiptenbrink/opaque-borink

In the end whatever may be picked, at least there would be an alternative.
